File: strategy/python/MazeGame.py
from abc import ABCMeta, abstractmethod
import random, time, sys

class MazeGame:

    def createMaze(self, mazeFactory):
        m = mazeFactory.makeMaze(5, 5)

        for x in range(5):
            for y in range(5):
                m.addComponent(mazeFactory.makeWall(), x, y)

        m.addComponent(mazeFactory.makeRoom(), 1, 2)
        m.addComponent(mazeFactory.makeRoom(), 2, 2)
        m.addComponent(mazeFactory.makeRoom(), 3, 2)
        m.addComponent(mazeFactory.makeRoom(), 1, 1)
        m.addComponent(mazeFactory.makeRoom(), 3, 1)
        m.addComponent(mazeFactory.makeRoom(), 1, 3)
        m.addComponent(mazeFactory.makeRoom(), 3, 3)

        return m

class MazeFactory:

    def makeMaze(self, x, y):
        return Maze(x, y)

    def makeRoom(self):
        return Room()

    def makeWall(self):
        return Wall()

class Maze:

    def __init__(self, x, y):
        self.length = x
        self.height = y
        self.grid = []
        
        for i in range(x):
            self.grid.append([])
            for j in range(y):
                self.grid[i].append(0)

    def addComponent(self, component, x, y):
        if isinstance(component, MazeComponent):
            try:
                self.grid[x][y] = component
            except:
                return

    def printMaze(self):
        for y in range(self.length):
            for x in range(self.height):
                print(self.grid[x][y].getSymbol(), end="")
            print("")

class MazeComponent:
    __metaclass__ = ABCMeta

    @abstractmethod
    def isSolid(self): pass

    @abstractmethod
    def getSymbol(self): pass

class Room(MazeComponent):
    def isSolid(self):
        return False

    def getSymbol(self):
        return '.'

class Wall(MazeComponent):
    def isSolid(self):
        return True

    def getSymbol(self):
        return '#'

class CoinFactory(MazeFactory):
    def makeRoom(self):
        return CoinRoom()

class CoinRoom(Room):
    def __init__(self):
        self.coin = False
        if random.randint(0, 100) >= 75:
            self.coin = True

    def getSymbol(self):
        if self.coin == True:
            return '@'
        else:
            return '.'

def main():
    mf = None
    if len(sys.argv) == 2:
        if sys.argv[1] == 'maze':
            mf = MazeFactory()
        elif sys.argv[1] == 'coin':
            mf = CoinFactory()
        else:
            print("Invalid input.")
            sys.exit()
    else:
        print("Usage: python MazeGame.py <maze|coin>")
        sys.exit()

    t1 = time.time()
    mg = MazeGame()
    m = mg.createMaze(mf)
    t2 = time.time() - t1
    print(t2)
    m.printMaze()

if __name__ == "__main__":
    main()

File: strategy/java/MazeGame.java
import java.util.Random;

public class MazeGame {

    public static Maze createMaze(MazeFactory mf) {
        Maze m = mf.makeMaze(5, 5);

        for (int x = 0; x < 5; x++) {
            for (int y = 0; y < 5; y++) {
                m.addComponent(mf.makeWall(), x, y);
            }
        }

        m.addComponent(mf.makeRoom(), 1, 2);
        m.addComponent(mf.makeRoom(), 2, 2);
        m.addComponent(mf.makeRoom(), 3, 2);
        m.addComponent(mf.makeRoom(), 1, 1);
        m.addComponent(mf.makeRoom(), 3, 1);
        m.addComponent(mf.makeRoom(), 1, 3);
        m.addComponent(mf.makeRoom(), 3, 3);

        return m;
    }

    public static void main(String[] args) {
        MazeFactory mf = null;
        if (args.length > 0) {
            if (args[0].equals("maze")) {
                mf = new MazeFactory();
            } else if (args[0].equals("coin")) {
                mf = new CoinFactory();
            } else {
                
            }
        } else {
            System.out.println("Usage: java MazeGame <coin|maze>");
            System.exit(0);
        }
        Maze m = createMaze(mf);
        m.printMaze();

    }
}

class MazeFactory {

    Maze makeMaze(int x, int y) {
        return new Maze(x, y);
    }
    
    Room makeRoom() {
        return new Room();
    }

    Wall makeWall() {
        return new Wall();
    }

}

class CoinFactory extends MazeFactory {
    
    Room makeRoom() {
        return new CoinRoom();
    }
}

class Maze {

    private MazeComponent[][] map;
    private int length;
    private int height;

    public Maze(int x, int y) {
        map = new MazeComponent[x][y];
        length = x;
        height = y;
    }

    public void addComponent(MazeComponent c, int x, int y) {
        if (!(x >= length || y >= height || x < 0 || y < 0)) {
            map[x][y] = c;
        }
    }

    public void printMaze() {
        for (int y = 0; y < 5; y++) {
            for (int x = 0; x < 5; x++) {
                System.out.print(map[x][y].getSymbol());
            }
            System.out.println("");
        }
    }
}

class Room implements MazeComponent {
    
    public boolean isSolid() {
        return false;
    }

    public char getSymbol() {
        return '.';
    }
}

class Wall implements MazeComponent {
    
    public boolean isSolid() {
        return true;
    }

    public char getSymbol() {
        return '#';
    }
}

class CoinRoom extends Room {
    boolean coin;

    public CoinRoom() {
        Random r = new Random();
        if (r.nextInt(100) >= 75) {
            coin = true;
        }
    }

    public char getSymbol() {
        return coin ? '@' : '.';
    }
}
